vault/seal_autoseal.go

```diff
@@ -499,3 +509,67 @@ func (d *autoSeal) migrateRecoveryConfig(ctx context.Context) error {
 
 	return nil
 }
+
+// StartHealthCheck starts a goroutine that tests the health of the auto-unseal backend once every 10 minutes.
+// If unhealthy, logs a warning on the condition and begins testing every one minute until healthy again.
+func (d *autoSeal) StartHealthCheck() {
+	d.healthCheck = time.NewTicker(sealHeathTestIntervalNominal)
+	d.healthCheckStop = make(chan struct{})
+	go func() {
+		lastTestOk := true
+		lastSeenOk := time.Now()
+		for {
+			select {
+			case <-d.healthCheckStop:
+				if d.healthCheck != nil {
+					d.healthCheck.Stop()
+				}
+				d.healthCheck = nil
+				d.healthCheckStop = nil
+				return
+			case t := <-d.healthCheck.C:
+				testVal := fmt.Sprintf("Heartbeat %d", mathrand.Intn(1000))
+				ciphertext, err := d.Wrapper.Encrypt(d.core.activeContext, []byte(testVal), nil)
+				if err != nil {
+					d.logger.Warn("failed to encrypt seal health test value, seal backend may be unreachable", "error", err)
+					if lastTestOk {
+						d.healthCheck.Reset(sealHeathTestIntervalUnhealthy)
+					}
+					lastTestOk = false
+				} else {
+					plaintext, err := d.Wrapper.Decrypt(d.core.activeContext, ciphertext, nil)
+					if err != nil {
+						d.logger.Warn("failed to decrypt seal health test value, seal backend may be unreachable", "error", err)
+						if lastTestOk {
+							d.healthCheck.Reset(sealHeathTestIntervalUnhealthy)
+						}
+						lastTestOk = false
+					}
+					if !bytes.Equal([]byte(testVal), plaintext) {
+						d.logger.Warn("seal health test value failed to decrypt to expected value")
+						if lastTestOk {
+							d.healthCheck.Reset(sealHeathTestIntervalUnhealthy)
+						}
+						lastTestOk = false
+					} else {
+						d.logger.Debug("seal health test passed")
+						if !lastTestOk {
+							d.logger.Info("seal backend is now healthy again", "downtime", t.Sub(lastSeenOk).String())
+						}
+						if !lastTestOk {
+							d.healthCheck.Reset(sealHeathTestIntervalNominal)
+						}
+						lastTestOk = true
+						lastSeenOk = t
+					}
+				}
+			}
+		}
+	}()
+}
+
+func (d *autoSeal) StopHealthCheck() {
+	if d.healthCheckStop != nil {
+		close(d.healthCheckStop)
```

Comment:
Maybe assign nil to the field so that if for some reason we get called twice we don't panic?

Reply:
Yep.